File: Typescript/src/DelegationSortedIntList.ts
/**
 * DelegationSortedIntList -- a variant of a SortedIntList that keeps
 * count of the number of attempted element insertions (not to be confused
 * with the current size, which goes down when an element is removed)
 * and exports an accessor (totalAdded) for this count.
 *
 * @author Nora Shoemaker
 *
 */
import {SortedIntList} from "./hidden/SortedIntListLibrary";
import {IntegerList} from "./IntegerList";

class DelegationSortedIntList implements IntegerList {

  // The SortedIntList object that this class delegates to.
  private sortedIntList: SortedIntList;

  // The total number of elements added to the list.
  private totalAdded: number;

  /**
   * Construct a new DelegationSortedIntList object.
   */
  constructor() {
    this.sortedIntList = new SortedIntList();
    this.totalAdded = 0;
  }

  /**
   * Add a number to the list.
   * @param num the number to add to the list
   * @return true if the number was added to the list, false otherwise
   */
  add(num: number): boolean {
    if (this.sortedIntList.add(num)) {
      this.totalAdded++;
      return true;
    }
    return false;
  }

  /**
   * Add all numbers from the given list to this list.
   * @param list the list of numbers to add to this list
   * @return true if all numbers were added to the list, false otherwise
   */
  addAll(list: IntegerList): boolean {
    if (this.sortedIntList.addAll(list)) {
      this.totalAdded += list.size();
      return true;
    }
    return false;
  }

  /**
   * Get the number at the given index.
   * @param index the index of the number to get
   * @return the number at the given index
   */
  get(index: number): number {
    return this.sortedIntList.get(index);
  }

  /**
   * Remove the number at the given index.
   * @param num the number to remove
   * @return true if the number was removed, false otherwise
   */
  remove(num: number): boolean {
    return this.sortedIntList.remove(num);
  }


  /**
   * Remove all numbers from the given list from this list.
   * @param list the list of numbers to remove from this list
   * @return true if all numbers were removed from the list, false otherwise
   */
  removeAll(list: IntegerList): boolean {
    return this.sortedIntList.removeAll(list);
  }

  /**
   * Get the size of the list.
   * @return the size of the list
   */
  size(): number {
    return this.sortedIntList.size();
  }

  /**
   * Get the total number of elements added to the list.
   * @return the total number of elements added to the list
   */
  getTotalAdded(): number {
    return this.totalAdded;
  }

}

export {DelegationSortedIntList}

File: Java/src/main/java/edu/cmu/cs/cs214/rec04/DelegationSortedIntList.java
package edu.cmu.cs.cs214.rec04;

/**
 * DelegationSortedIntList -- a variant of a SortedIntList that keeps
 * count of the number of attempted element insertions (not to be confused
 * with the current size, which goes down when an element is removed)
 * and exports an accessor (totalAdded) for this count.
 *
 * @author Nora Shoemaker
 */

public class DelegationSortedIntList implements IntegerList {

    /**
     * sortedIntList -- a SortedIntList object to delegate to
     */
    private final SortedIntList sortedIntList;

    /**
     * totalAdded -- the total number of elements added to the list
     */
    private int totalAdded;

    /**
     * Constructor for the DelegationSortedIntList class
     */
    public DelegationSortedIntList() {
        this.sortedIntList = new SortedIntList();
        this.totalAdded = 0;
    }

    /**
     * Delegate the constructor to a SortedIntList object
     *
     * @param num an integer to be added to the list
     * @return true if the integer was added to the list, false otherwise
     */
    @Override
    public boolean add(int num) {
        if (this.sortedIntList.add(num)) {
            this.totalAdded++;
            return true;
        }
        return false;
    }

    /**
     * Delegate the constructor to a SortedIntList object
     *
     * @param list IntegerList containing elements to be added to the list
     * @return true if all elements were added to the list, false otherwise
     */
    @Override
    public boolean addAll(IntegerList list) {
        if (this.sortedIntList.addAll(list)) {
            this.totalAdded += list.size();
            return true;
        }
        return false;
    }

    /**
     * Delegate the constructor to a SortedIntList object
     *
     * @param index index of the element to return
     * @return the element at the specified position in this list
     */
    @Override
    public int get(int index) {
        return this.sortedIntList.get(index);
    }

    /**
     * Delegate the constructor to a SortedIntList object
     *
     * @param num an integer to be added to the list
     * @return the index of the element in the list
     */
    @Override
    public boolean remove(int num) {
        return this.sortedIntList.remove(num);
    }

    /**
     * Delegate the constructor to a SortedIntList object
     *
     * @param list IntegerList containing elements to be removed from the list
     * @return true if all elements were removed from the list, false otherwise
     */
    @Override
    public boolean removeAll(IntegerList list) {
        return this.sortedIntList.removeAll(list);
    }

    /**
     * Delegate the constructor to a SortedIntList object
     *
     * @return the number of elements in the list
     */
    @Override
    public int size() {
        return this.sortedIntList.size();
    }

    /**
     * Get the total number of elements added to the list
     *
     * @return the total number of elements added to the list
     */
    public int getTotalAdded() {
        return this.totalAdded;
    }

}